File: src/shell/builtin/touch.zig
opts: Opts = .{},
state: union(enum) {
    idle,
    exec: struct {
        started: bool = false,
        tasks_count: usize = 0,
        tasks_done: usize = 0,
        output_done: usize = 0,
        output_waiting: usize = 0,
        started_output_queue: bool = false,
        args: []const [*:0]const u8,
        err: ?JSC.SystemError = null,
    },
    waiting_write_err,
    done,
} = .idle,

pub fn format(this: *const Touch, comptime fmt: []const u8, opts: std.fmt.FormatOptions, writer: anytype) !void {
    _ = fmt; // autofix
    _ = opts; // autofix
    try writer.print("Touch(0x{x}, state={s})", .{ @intFromPtr(this), @tagName(this.state) });
}

pub fn deinit(this: *Touch) void {
    log("{} deinit", .{this});
}

pub fn start(this: *Touch) Yield {
    const filepath_args = switch (this.opts.parse(this.bltn().argsSlice())) {
        .ok => |filepath_args| filepath_args,
        .err => |e| {
            const buf = switch (e) {
                .illegal_option => |opt_str| this.bltn().fmtErrorArena(.touch, "illegal option -- {s}\n", .{opt_str}),
                .show_usage => Builtin.Kind.touch.usageString(),
                .unsupported => |unsupported| this.bltn().fmtErrorArena(.touch, "unsupported option, please open a GitHub issue -- {s}\n", .{unsupported}),
            };

            return this.writeFailingError(buf, 1);
        },
    } orelse {
        return this.writeFailingError(Builtin.Kind.touch.usageString(), 1);
    };

    this.state = .{
        .exec = .{
            .args = filepath_args,
        },
    };

    return this.next();
}

pub fn next(this: *Touch) Yield {
    switch (this.state) {
        .idle => @panic("Invalid state"),
        .exec => {
            var exec = &this.state.exec;
            if (exec.started) {
                if (this.state.exec.tasks_done >= this.state.exec.tasks_count and this.state.exec.output_done >= this.state.exec.output_waiting) {
                    const exit_code: ExitCode = if (this.state.exec.err != null) 1 else 0;
                    this.state = .done;
                    return this.bltn().done(exit_code);
                }
                return .suspended;
            }

            exec.started = true;
            exec.tasks_count = exec.args.len;

            for (exec.args) |dir_to_mk_| {
                const dir_to_mk = dir_to_mk_[0..std.mem.len(dir_to_mk_) :0];
                var task = ShellTouchTask.create(this, this.opts, dir_to_mk, this.bltn().parentCmd().base.shell.cwdZ());
                task.schedule();
            }
            return .suspended;
        },
        .waiting_write_err => return .failed,
        .done => return this.bltn().done(0),
    }
}

pub fn onIOWriterChunk(this: *Touch, _: usize, e: ?JSC.SystemError) Yield {
    if (this.state == .waiting_write_err) {
        return this.bltn().done(1);
    }

    if (e) |err| err.deref();

    return this.next();
}

pub fn writeFailingError(this: *Touch, buf: []const u8, exit_code: ExitCode) Yield {
    if (this.bltn().stderr.needsIO()) |safeguard| {
        this.state = .waiting_write_err;
        return this.bltn().stderr.enqueue(this, buf, safeguard);
    }

    _ = this.bltn().writeNoIO(.stderr, buf);

    return this.bltn().done(exit_code);
}

pub fn onShellTouchTaskDone(this: *Touch, task: *ShellTouchTask) void {
    log("{} onShellTouchTaskDone {} tasks_done={d} tasks_count={d}", .{ this, task, this.state.exec.tasks_done, this.state.exec.tasks_count });

    defer bun.default_allocator.destroy(task);
    this.state.exec.tasks_done += 1;
    const err = task.err;

    if (err) |e| {
        const output_task: *ShellTouchOutputTask = bun.new(ShellTouchOutputTask, .{
            .parent = this,
            .output = .{ .arrlist = .{} },
            .state = .waiting_write_err,
        });
        const error_string = this.bltn().taskErrorToString(.touch, e);
        this.state.exec.err = e;
        output_task.start(error_string).run();
        return;
    }

    this.next().run();
}

pub const ShellTouchOutputTask = OutputTask(Touch, .{
    .writeErr = ShellTouchOutputTaskVTable.writeErr,
    .onWriteErr = ShellTouchOutputTaskVTable.onWriteErr,
    .writeOut = ShellTouchOutputTaskVTable.writeOut,
    .onWriteOut = ShellTouchOutputTaskVTable.onWriteOut,
    .onDone = ShellTouchOutputTaskVTable.onDone,
});

const ShellTouchOutputTaskVTable = struct {
    pub fn writeErr(this: *Touch, childptr: anytype, errbuf: []const u8) ?Yield {
        if (this.bltn().stderr.needsIO()) |safeguard| {
            this.state.exec.output_waiting += 1;
            return this.bltn().stderr.enqueue(childptr, errbuf, safeguard);
        }
        _ = this.bltn().writeNoIO(.stderr, errbuf);
        return null;
    }

    pub fn onWriteErr(this: *Touch) void {
        this.state.exec.output_done += 1;
    }

    pub fn writeOut(this: *Touch, childptr: anytype, output: *OutputSrc) ?Yield {
        if (this.bltn().stdout.needsIO()) |safeguard| {
            this.state.exec.output_waiting += 1;
            const slice = output.slice();
            log("THE SLICE: {d} {s}", .{ slice.len, slice });
            return this.bltn().stdout.enqueue(childptr, slice, safeguard);
        }
        _ = this.bltn().writeNoIO(.stdout, output.slice());
        return null;
    }

    pub fn onWriteOut(this: *Touch) void {
        this.state.exec.output_done += 1;
    }

    pub fn onDone(this: *Touch) Yield {
        return this.next();
    }
};

pub const ShellTouchTask = struct {
    touch: *Touch,

    opts: Opts,
    filepath: [:0]const u8,
    cwd_path: [:0]const u8,

    err: ?JSC.SystemError = null,
    task: JSC.WorkPoolTask = .{ .callback = &runFromThreadPool },
    event_loop: JSC.EventLoopHandle,
    concurrent_task: JSC.EventLoopTask,

    pub fn format(this: *const ShellTouchTask, comptime fmt: []const u8, opts: std.fmt.FormatOptions, writer: anytype) !void {
        _ = fmt; // autofix
        _ = opts; // autofix
        try writer.print("ShellTouchTask(0x{x}, filepath={s})", .{ @intFromPtr(this), this.filepath });
    }

    pub fn deinit(this: *ShellTouchTask) void {
        if (this.err) |*e| {
            e.deref();
        }
        bun.default_allocator.destroy(this);
    }

    pub fn create(touch: *Touch, opts: Opts, filepath: [:0]const u8, cwd_path: [:0]const u8) *ShellTouchTask {
        const task = bun.default_allocator.create(ShellTouchTask) catch bun.outOfMemory();
        task.* = ShellTouchTask{
            .touch = touch,
            .opts = opts,
            .cwd_path = cwd_path,
            .filepath = filepath,
            .event_loop = touch.bltn().eventLoop(),
            .concurrent_task = JSC.EventLoopTask.fromEventLoop(touch.bltn().eventLoop()),
        };
        return task;
    }

    pub fn schedule(this: *@This()) void {
        debug("{} schedule", .{this});
        WorkPool.schedule(&this.task);
    }

    pub fn runFromMainThread(this: *@This()) void {
        debug("{} runFromJS", .{this});
        this.touch.onShellTouchTaskDone(this);
    }

    pub fn runFromMainThreadMini(this: *@This(), _: *void) void {
        this.runFromMainThread();
    }

    fn runFromThreadPool(task: *JSC.WorkPoolTask) void {
        var this: *ShellTouchTask = @fieldParentPtr("task", task);
        debug("{} runFromThreadPool", .{this});

        // We have to give an absolute path
        const filepath: [:0]const u8 = brk: {
            if (ResolvePath.Platform.auto.isAbsolute(this.filepath)) break :brk this.filepath;
            const parts: []const []const u8 = &.{
                this.cwd_path[0..],
                this.filepath[0..],
            };
            break :brk ResolvePath.joinZ(parts, .auto);
        };

        var node_fs = JSC.Node.fs.NodeFS{};
        const milliseconds: f64 = @floatFromInt(std.time.milliTimestamp());
        const atime: JSC.Node.TimeLike = if (bun.Environment.isWindows) milliseconds / 1000.0 else JSC.Node.TimeLike{
            .sec = @intFromFloat(@divFloor(milliseconds, std.time.ms_per_s)),
            .nsec = @intFromFloat(@mod(milliseconds, std.time.ms_per_s) * std.time.ns_per_ms),
        };
        const mtime = atime;
        const args = JSC.Node.fs.Arguments.Utimes{
            .atime = atime,
            .mtime = mtime,
            .path = .{ .string = bun.PathString.init(filepath) },
        };
        if (node_fs.utimes(args, .sync).asErr()) |err| out: {
            if (err.getErrno() == .NOENT) {
                const perm = 0o664;
                switch (Syscall.open(filepath, bun.O.CREAT | bun.O.WRONLY, perm)) {
                    .result => |fd| {
                        fd.close();
                        break :out;
                    },
                    .err => |e| {
                        this.err = e.withPath(bun.default_allocator.dupe(u8, filepath) catch bun.outOfMemory()).toShellSystemError();
                        break :out;
                    },
                }
            }
            this.err = err.withPath(bun.default_allocator.dupe(u8, filepath) catch bun.outOfMemory()).toShellSystemError();
        }

        if (this.event_loop == .js) {
            this.event_loop.js.enqueueTaskConcurrent(this.concurrent_task.js.from(this, .manual_deinit));
        } else {
            this.event_loop.mini.enqueueTaskConcurrent(this.concurrent_task.mini.from(this, "runFromMainThreadMini"));
        }
    }
};

const Opts = struct {
    /// -a
    ///
    /// change only the access time
    access_time_only: bool = false,

    /// -c, --no-create
    ///
    /// do not create any files
    no_create: bool = false,

    /// -d, --date=STRING
    ///
    /// parse STRING and use it instead of current time
    date: ?[]const u8 = null,

    /// -h, --no-dereference
    ///
    /// affect each symbolic link instead of any referenced file
    /// (useful only on systems that can change the timestamps of a symlink)
    no_dereference: bool = false,

    /// -m
    ///
    /// change only the modification time
    modification_time_only: bool = false,

    /// -r, --reference=FILE
    ///
    /// use this file's times instead of current time
    reference: ?[]const u8 = null,

    /// -t STAMP
    ///
    /// use [[CC]YY]MMDDhhmm[.ss] instead of current time
    timestamp: ?[]const u8 = null,

    /// --time=WORD
    ///
    /// change the specified time:
    /// WORD is access, atime, or use: equivalent to -a
    /// WORD is modify or mtime: equivalent to -m
    time: ?[]const u8 = null,

    const Parse = FlagParser(*@This());

    pub fn parse(opts: *Opts, args: []const [*:0]const u8) Result(?[]const [*:0]const u8, ParseError) {
        return Parse.parseFlags(opts, args);
    }

    pub fn parseLong(this: *Opts, flag: []const u8) ?ParseFlagResult {
        _ = this;
        if (bun.strings.eqlComptime(flag, "--no-create")) {
            return .{
                .unsupported = unsupportedFlag("--no-create"),
            };
        }

        if (bun.strings.eqlComptime(flag, "--date")) {
            return .{
                .unsupported = unsupportedFlag("--date"),
            };
        }

        if (bun.strings.eqlComptime(flag, "--reference")) {
            return .{
                .unsupported = unsupportedFlag("--reference=FILE"),
            };
        }

        if (bun.strings.eqlComptime(flag, "--time")) {
            return .{
                .unsupported = unsupportedFlag("--reference=FILE"),
            };
        }

        return null;
    }

    pub fn parseShort(this: *Opts, char: u8, smallflags: []const u8, i: usize) ?ParseFlagResult {
        _ = this;
        switch (char) {
            'a' => {
                return .{ .unsupported = unsupportedFlag("-a") };
            },
            'c' => {
                return .{ .unsupported = unsupportedFlag("-c") };
            },
            'd' => {
                return .{ .unsupported = unsupportedFlag("-d") };
            },
            'h' => {
                return .{ .unsupported = unsupportedFlag("-h") };
            },
            'm' => {
                return .{ .unsupported = unsupportedFlag("-m") };
            },
            'r' => {
                return .{ .unsupported = unsupportedFlag("-r") };
            },
            't' => {
                return .{ .unsupported = unsupportedFlag("-t") };
            },
            else => {
                return .{ .illegal_option = smallflags[1 + i ..] };
            },
        }

        return null;
    }
};

pub inline fn bltn(this: *Touch) *Builtin {
    const impl: *Builtin.Impl = @alignCast(@fieldParentPtr("touch", this));
    return @fieldParentPtr("impl", impl);
}

// --
const debug = bun.Output.scoped(.ShellTouch, true);
const Touch = @This();
const log = debug;
const std = @import("std");
const bun = @import("bun");
const Yield = bun.shell.Yield;
const shell = bun.shell;
const ExitCode = shell.ExitCode;
const JSC = bun.JSC;
const WorkPool = bun.JSC.WorkPool;
const ResolvePath = bun.path;
const Syscall = bun.sys;
const interpreter = @import("../interpreter.zig");
const Interpreter = interpreter.Interpreter;
const Builtin = Interpreter.Builtin;
const Result = Interpreter.Builtin.Result;
const ParseError = interpreter.ParseError;
const ParseFlagResult = interpreter.ParseFlagResult;
const FlagParser = interpreter.FlagParser;
const unsupportedFlag = interpreter.unsupportedFlag;
const OutputTask = interpreter.OutputTask;
const OutputSrc = interpreter.OutputSrc;

File: src/runtime/shell/builtin/touch.rs
use crate::shell::ExitCode;
use crate::shell::builtin::{Builtin, BuiltinState, Kind};
use crate::shell::interpreter::{
    FlagParser, Interpreter, NodeId, OutputQueue, OutputSrc, OutputTask, OutputTaskVTable,
    ParseFlagResult, ShellTask, unsupported_flag,
};
use crate::shell::yield_::Yield;
use bun_ptr::JsCellRefMut;

#[derive(Default)]
pub struct Touch {
    pub(crate) state: State,
}

#[derive(Default)]
pub enum State {
    #[default]
    Idle,
    Exec(ExecState),
    WaitingWriteErr,
    Done,
}

pub struct ExecState {
    pub(crate) started: bool,
    pub(crate) tasks_count: usize,
    pub(crate) tasks_done: usize,
    /// Index into argv where filepath args start.
    pub(crate) args_start: usize,
    pub(crate) err: Option<bun_sys::Error>,
    pub(crate) output: OutputQueue<Touch>,
}

impl Touch {
    pub(crate) fn start(interp: &Interpreter, cmd: NodeId) -> Yield {
        let mut opts = Opts::default();
        let args_start = {
            match Builtin::parse_flags(interp, cmd, &mut opts) {
                Ok(Some(start)) => start,
                Ok(None) => {
                    Self::state_mut(interp, cmd).state = State::WaitingWriteErr;
                    return Builtin::write_failing_error(
                        interp,
                        cmd,
                        Kind::Touch.usage_string(),
                        1,
                    );
                }
                Err(e) => {
                    return Builtin::fail_parse(interp, cmd, Kind::Touch, &e, || {
                        Self::state_mut(interp, cmd).state = State::WaitingWriteErr
                    });
                }
            }
        };
        Self::state_mut(interp, cmd).state = State::Exec(ExecState {
            started: false,
            tasks_count: 0,
            tasks_done: 0,
            args_start,
            err: None,
            output: OutputQueue::default(),
        });
        Self::next(interp, cmd)
    }

    fn next(interp: &Interpreter, cmd: NodeId) -> Yield {
        enum Action {
            Done(ExitCode),
            Schedule(usize),
            Suspend,
            Failed,
            AlreadyDone,
        }
        let action = match &mut Self::state_mut(interp, cmd).state {
            State::Idle => panic!("Invalid state"),
            State::Exec(exec) => {
                if exec.started {
                    if exec.tasks_done >= exec.tasks_count && exec.output.drained() {
                        let code: ExitCode = if exec.err.is_some() { 1 } else { 0 };
                        exec.err = None;
                        Action::Done(code)
                    } else {
                        Action::Suspend
                    }
                } else {
                    exec.started = true;
                    Action::Schedule(exec.args_start)
                }
            }
            State::WaitingWriteErr => Action::Failed,
            State::Done => Action::AlreadyDone,
        };
        match action {
            Action::Suspend => Yield::suspended(),
            Action::Failed => Yield::failed(),
            Action::AlreadyDone => Builtin::done(interp, cmd, 0),
            Action::Done(code) => {
                Self::state_mut(interp, cmd).state = State::Done;
                Builtin::done(interp, cmd, code)
            }
            Action::Schedule(args_start) => {
                let argc = Builtin::argc(interp, cmd);
                if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                    exec.tasks_count = argc - args_start;
                }
                let cwd = Builtin::shell(interp, cmd).borrow().cwd().to_vec();
                for i in args_start..argc {
                    let path = Builtin::of(interp, cmd).arg_bytes(i).to_vec();
                    let task = ShellTouchTask::create(cmd, path, cwd.clone(), interp);
                    ShellTask::schedule(task);
                }
                Yield::suspended()
            }
        }
    }

    pub(crate) fn on_io_writer_chunk(
        interp: &Interpreter,
        cmd: NodeId,
        written: usize,
        e: Option<bun_sys::SystemError>,
    ) -> Yield {
        // Only the usage error is written directly; task output goes through
        // `OutputTask::on_chunk`.
        let _ = (written, e);
        debug_assert!(matches!(
            Self::state_mut(interp, cmd).state,
            State::WaitingWriteErr
        ));
        Builtin::done(interp, cmd, 1)
    }

    fn on_shell_touch_task_done(interp: &Interpreter, cmd: NodeId, mut task: ShellTouchTask) {
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.tasks_done += 1;
        }
        if let Some(e) = task.err.take() {
            let output_task = OutputTask::<Touch>::new(cmd, OutputSrc::Arrlist(Vec::new()));
            let errstr = Builtin::task_error_to_string(Kind::Touch, &e);
            if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                exec.err = Some(e);
            }
            OutputTask::<Touch>::start(output_task, interp, Some(&errstr)).run(interp);
            return;
        }
        Self::next(interp, cmd).run(interp);
    }
}

impl OutputTaskVTable for Touch {
    fn output_queue(interp: &Interpreter, cmd: NodeId) -> JsCellRefMut<'_, OutputQueue<Self>> {
        JsCellRefMut::map(Self::state_mut(interp, cmd), |me| match &mut me.state {
            State::Exec(exec) => &mut exec.output,
            _ => unreachable!("touch output outside Exec"),
        })
    }

    fn on_done(interp: &Interpreter, cmd: NodeId) -> Yield {
        Self::next(interp, cmd)
    }
}

/// utimes() the path (creating it on ENOENT) on a worker thread.
pub struct ShellTouchTask {
    pub(crate) cmd: NodeId,
    pub(crate) filepath: Vec<u8>,
    pub(crate) cwd_path: Vec<u8>,
    pub(crate) err: Option<bun_sys::Error>,
    pub task: ShellTask,
}

crate::shell_task!(ShellTouchTask);

impl ShellTouchTask {
    pub(crate) fn create(
        cmd: NodeId,
        filepath: Vec<u8>,
        cwd_path: Vec<u8>,
        interp: &Interpreter,
    ) -> Box<ShellTouchTask> {
        Box::new(ShellTouchTask {
            cmd,
            filepath,
            cwd_path,
            err: None,
            task: ShellTask::new(interp),
        })
    }

    /// utimes() the path; on ENOENT
    /// fall back to `open(O_CREAT|O_WRONLY, 0o664)`.
    pub(crate) fn run_from_thread_pool(this: &mut ShellTouchTask) {
        use bun_paths::resolve_path::{self, Platform, platform};
        use bun_sys::FdExt as _;
        // We have to give an absolute path. An operand that does not fit the
        // path buffer is still passed on whole, so the OS reports ENAMETOOLONG
        // for it like for any other operand.
        let mut spill = Vec::new();
        let filepath: &bun_core::ZStr = if Platform::AUTO.is_absolute(&this.filepath) {
            // Re-terminate (`filepath` is the bare argv bytes without the
            // trailing NUL).
            resolve_path::join_z_spill::<platform::Auto>(&mut spill, &[&this.filepath])
        } else {
            resolve_path::join_z_spill::<platform::Auto>(
                &mut spill,
                &[&this.cwd_path, &this.filepath],
            )
        };

        // Call the bun_sys layer directly (uv_fs_utime on Windows) to avoid
        // the heavyweight NodeFS state.
        let milliseconds = bun_core::time::milli_timestamp();
        let atime = bun_sys::TimeLike {
            sec: milliseconds.div_euclid(1_000),
            nsec: milliseconds.rem_euclid(1_000) * 1_000_000,
        };
        let mtime = atime;
        if let Err(err) = bun_sys::utimens(filepath, atime, mtime) {
            'out: {
                if err.get_errno() == bun_sys::E::ENOENT {
                    const PERM: bun_sys::Mode = 0o664;
                    match bun_sys::open(filepath, bun_sys::O::CREAT | bun_sys::O::WRONLY, PERM) {
                        Ok(fd) => {
                            fd.close();
                            break 'out;
                        }
                        Err(e) => {
                            this.err = Some(e.with_path(filepath.as_bytes()));
                            break 'out;
                        }
                    }
                }
                this.err = Some(err.with_path(filepath.as_bytes()));
            }
        }
        // Worker→main bounce-back is posted by `ShellTask::run_owned` after
        // this returns.
    }
}

// `runtime::dispatch::run_task`'s `task_tag::ShellTouchTask` arm reboxes the
// pointer `ShellTask::on_finish` posted; a completion that will not run drops
// the keep-alive and the box.

impl crate::shell::interpreter::ShellTaskCtx for ShellTouchTask {
    fn shell_task(&self) -> &ShellTask {
        &self.task
    }
    fn shell_task_mut(&mut self) -> &mut ShellTask {
        &mut self.task
    }
    fn run_from_thread_pool(&mut self) {
        Self::run_from_thread_pool(self)
    }
    fn run_from_main_thread(self: Box<Self>, interp: &Interpreter) {
        let cmd = self.cmd;
        Touch::on_shell_touch_task_done(interp, cmd, *self);
    }
}

#[derive(Clone, Copy, Default)]
pub struct Opts {}

impl FlagParser for Opts {
    fn parse_long(&mut self, flag: &[u8]) -> Option<ParseFlagResult> {
        match flag {
            b"--no-create" => Some(ParseFlagResult::Unsupported(unsupported_flag(
                b"--no-create",
            ))),
            b"--date" => Some(ParseFlagResult::Unsupported(unsupported_flag(b"--date"))),
            b"--reference" => Some(ParseFlagResult::Unsupported(unsupported_flag(
                b"--reference=FILE",
            ))),
            b"--time" => Some(ParseFlagResult::Unsupported(unsupported_flag(
                b"--reference=FILE",
            ))),
            _ => None,
        }
    }

    fn parse_short(&mut self, ch: u8, smallflags: &[u8], i: usize) -> Option<ParseFlagResult> {
        match ch {
            b'a' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-a"))),
            b'c' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-c"))),
            b'd' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-d"))),
            b'h' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-h"))),
            b'm' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-m"))),
            b'r' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-r"))),
            b't' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-t"))),
            _ => Some(ParseFlagResult::IllegalOption(smallflags[1 + i..].into())),
        }
    }
}
